pymtl3/stdlib/rtl/queues.py

```diff
@@ -86,11 +86,298 @@ def construct( s, num_entries=2 ):
     @s.update
     def up_rdy_signals():
       if ~s.reset:
-        s.enq_rdy = s.count < s.num_entries
-        s.deq_rdy = s.count > 0
+        s.enq_rdy = b1(1) if s.count < s.num_entries else b1(0)
+        s.deq_rdy = b1(1) if s.count > 0 else b1(0)
       else:
-        s.enq_rdy = Bits1(0)
-        s.deq_rdy = Bits1(0)
+        s.enq_rdy = b1(0)
+        s.deq_rdy = b1(0)
+
+    @s.update
+    def up_xfer_signals():
+      s.enq_xfer = s.enq_en and s.enq_rdy
+      s.deq_xfer = s.deq_en and s.deq_rdy
+
+    @s.update
+    def up_next():
+      s.head_next = s.head - 1 if s.head > 0 else PtrType( s.last_idx )
+      s.tail_next = s.tail + 1 if s.tail < s.last_idx else PtrType(0)
+
+    @s.update_on_edge
+    def up_reg():
+
+      if s.reset:
+        s.head  = PtrType(0)
+        s.tail  = PtrType(0)
+        s.count = CountType(0)
+
+      else:
+        s.head  = s.head_next if s.deq_xfer else s.head
+        s.tail  = s.tail_next if s.enq_xfer else s.tail
+        s.count = s.count + b1(1) if s.enq_xfer and not s.deq_xfer else \
+                  s.count - b1(1) if s.deq_xfer and not s.enq_xfer else \
+                  s.count
+
+#-------------------------------------------------------------------------
+# NormalQueueRTL
+#-------------------------------------------------------------------------
+
+class NormalQueueRTL( Component ):
+
+  def construct( s, EntryType, num_entries=2 ):
+
+    # Interface
+
+    s.enq   = EnqIfcRTL( EntryType )
+    s.deq   = DeqIfcRTL( EntryType )
+    s.count = OutPort( mk_bits( clog2( num_entries ) ) )
+
+    # Components
+
+    s.ctrl  = NormalQueueCtrlRTL ( num_entries )
+    s.dpath = NormalQueueDpathRTL( EntryType, num_entries )
+
+    # Connect ctrl to data path
+
+    s.connect( s.ctrl.wen,     s.dpath.wen     )
+    s.connect( s.ctrl.waddr,   s.dpath.waddr   )
+    s.connect( s.ctrl.raddr,   s.dpath.raddr   )
+
+    # Connect to interface
+
+    s.connect( s.enq.en,  s.ctrl.enq_en   )
+    s.connect( s.enq.rdy, s.ctrl.enq_rdy  )
+    s.connect( s.deq.en,  s.ctrl.deq_en   )
+    s.connect( s.deq.rdy, s.ctrl.deq_rdy  )
+    s.connect( s.count,   s.ctrl.count    )
```

Comment:
Just found a bug here. Bit width of `s.count` and `s.ctrl.count` does not match. Why isn't `s.connect` complaining? @jsn1993

Reply:
Oh I see, sometimes their bit width can be the same. Our tests should test more configurations.